### PR TITLE
 db: introduce UsableSize flag in block volume Info

### PR DESCRIPTION
**Pull request by pkalever**
### What does this PR achieve? Why do we need it?

We should use newly expanded size on the block volume by growing filesystem on
the block device only when block volume is successfully expanded.

On partial failures (few nodes configured to size X and few nodes with size Y)
we should never proceed further for using the newly expanded size. This
might get us into block storage level inconsistencies.

To indicate the real usable size of a block volume, we have come up with
adding a new field to the block volume info.

#### Example 1:
$ heketi-cli blockvolume expand 95aaa622f466bc4ae9e5d460ac55505b --new-size=5
Name: blockvol_95aaa622f466bc4ae9e5d460ac55505b
Size: 5
UsableSize: 5
Volume Id: 95aaa622f466bc4ae9e5d460ac55505b
Cluster Id: 71d6c0a8f6e88c1ddc84b78b464aafd9
Hosts: [192.168.121.55 192.168.121.206 192.168.121.253]
IQN: iqn.2016-12.org.gluster-block:3fe22613-c669-4b04-ab6b-de279c84ca77
LUN: 0
Hacount: 3
Username:
Password:
Block Hosting Volume: b0cb4b999f9d1d386b093c81480e3964

$ heketi-cli blockvolume info 95aaa622f466bc4ae9e5d460ac55505b
Name: blockvol_95aaa622f466bc4ae9e5d460ac55505b
Size: 5
UsableSize: 5
Volume Id: 95aaa622f466bc4ae9e5d460ac55505b
Cluster Id: 71d6c0a8f6e88c1ddc84b78b464aafd9
Hosts: [192.168.121.55 192.168.121.206 192.168.121.253]
IQN: iqn.2016-12.org.gluster-block:3fe22613-c669-4b04-ab6b-de279c84ca77
LUN: 0
Hacount: 3
Username:
Password:
Block Hosting Volume: b0cb4b999f9d1d386b093c81480e3964

Here 'Size' matches 'UsableSize', hence we can proceed further to grow fs on
the device and use the 5GiB Space completely.

#### Example 2:

$ heketi-cli blockvolume expand 95aaa622f466bc4ae9e5d460ac55505b --new-size=10
Error: block volume resize failed: 192.168.121.55:[5.0 GiB] 192.168.121.206:[5.0 GiB], please check logs, and retry again.

$ heketi-cli blockvolume info 95aaa622f466bc4ae9e5d460ac55505b
Name: blockvol_95aaa622f466bc4ae9e5d460ac55505b
Size: 10
UsableSize: 5
Volume Id: 95aaa622f466bc4ae9e5d460ac55505b
Cluster Id: 71d6c0a8f6e88c1ddc84b78b464aafd9
Hosts: [192.168.121.55 192.168.121.206 192.168.121.253]
IQN: iqn.2016-12.org.gluster-block:3fe22613-c669-4b04-ab6b-de279c84ca77
LUN: 0
Hacount: 3
Username:
Password:
Block Hosting Volume: b0cb4b999f9d1d386b093c81480e3964

Here 'Size' doesn't match 'UsableSize', hence we shouldn't proceed further
to grow fs on the device, instead look at the logs, find and fix the real
problem and rerun the expand command again until we see a success i.e.
Size and UsableSize match.

Signed-off-by: Prasanna Kumar Kalever \<prasanna.kalever@redhat.com\>

### Notes for the reviewer
Supporting changes: https://github.com/gluster/gluster-block/pull/271

~~#### TODO:
Add tests for UsableSize Flag [Done now.]~~

### CODE REVIEW

**Comment by pkalever** [comment]
@phlogistonjohn @raghavendra-talur is this a known issue?

```
Ansible failed to complete successfully. Any error output should be
visible above. Please fix these errors and try again.
==> ERROR: unable to start test environment
==> FAILED TestSmokeTest
```

**Comment by raghavendra-talur** [comment]
> @phlogistonjohn @raghavendra-talur is this a known issue?
> 
> ```
> Ansible failed to complete successfully. Any error output should be
> visible above. Please fix these errors and try again.
> ==> ERROR: unable to start test environment
> ==> FAILED TestSmokeTest
> ```

I am not aware of any previous instance of this issue. @pkalever  Please set "set -x" in `tests/functional/run.sh` and refresh this patch.

**Comment by pkalever** [comment]
> I am not aware of any previous instance of this issue. @pkalever Please set "set -x" in `tests/functional/run.sh` and refresh this patch.

I will try this Talur. Thanks!

**Comment by pkalever** [comment]
> > I am not aware of any previous instance of this issue. @pkalever Please set "set -x" in `tests/functional/run.sh` and refresh this patch.
> 
> I will try this Talur. Thanks!

@raghavendra-talur we hit the issue again, but this time with 'set -x'. Please see if that rings a bell.

Thanks!

**Comment by pkalever** [comment]
I see it is failing in `TASK [gluster : install gluster packages]` resolving dependencies `Requires: glusterfs-libs = 4.1.9-1.el7\n           Available: glusterfs-libs-4.1.0-1.el7.x86_64`

**Comment by phlogistonjohn** [comment]
To get a better read of the failure I did:
```
curl -O 'https://ci.centos.org/job/gluster_heketi-functional/2360/consoleText'
sed -e 's/\\n/\n/g' consoleText  | less
```

It looks like its confused about what version of gluster to pull in for some reason. On first glance it seems to be mixing gluster 4.1.9 packages with gluster 6 packages. Needs more investigation.

**Comment by pkalever** [comment]
@phlogistonjohn @raghavendra-talur all checks have passed now, can you please help review this?

Thanks!

File: apps/glusterfs/operations_block_volume.go [hunk 1]
```diff
@@ -486,7 +488,10 @@ func (bve *BlockVolumeExpandOperation) Build() error {
 		}
 
 		requiredFreeSize := bve.newSize - bv.Info.Size
-		if bhv.Info.BlockInfo.FreeSize < requiredFreeSize {
+		if requiredFreeSize == 0 {
+			logger.Info("Attempting to fix failures seen by recent expand operation")
```
Comment (phlogistonjohn) [inline apps/glusterfs/operations_block_volume.go:492]: This log line would confuse me if I was debbuging things and saw it w/o looking at the code. I would 1. make this a warning to highlight that the state of the system is a tad bit out of whack and 2. reword it to clarify what is actually being done. Perhaps, "Re-executing block volume expansion on [%v]: volume not fully expanded on backend" where %v is to be replaced by the volume id.

File: executors/cmdexec/block_volume.go [hunk 2]
```diff
@@ -293,6 +294,26 @@ func (c *CmdExecutor) BlockVolumeInfo(host string, blockhostingvolume string,
 	blockVolumeInfo.Username = blockVolumeInfoExec.Gbid
 	blockVolumeInfo.Password = blockVolumeInfoExec.Password
 
+	// From gluster-block output,
+	// get the Minimum Size from the list of Resize failed nodes
+	var differSizes []int
+	for _, size := range blockVolumeInfoExec.ResizeFailed {
```
Comment (phlogistonjohn) [inline executors/cmdexec/block_volume.go:300]: Cant you do this in one loop? Pseudo-code:
```
minSize := -1
for _, size := range blockVolumeInfoExec.ResizeFailed {
    value, err := toGigaBytes(size)
    if err != nil {
        return err
    }
    if minSize == -1 || value < minSize {
        value = size {
    }
}
if minSIze != -1 {
    blockVolumeInfo.UsableSize = minSize
}
```

Also, please don't silently drop the error from toGigaBytes by just passing over it in the loop. If that function fails in production the best thing heketi can do is fail the operation noisily.

File: executors/cmdexec/block_volume.go [hunk 3]
```diff
@@ -337,8 +337,11 @@ func (s *CmdExecutor) BlockVolumeExpand(host string, blockHostingVolumeName stri
 			output)
 	} else if blockVolumeExpand.Result == "FAIL" {
 		// the fail flag was set in the output json
-		err = fmt.Errorf("Failed to expand block volume: %v",
-			blockVolumeExpand.ErrMsg)
+		if blockVolumeExpand.ErrMsg != "" {
+			err = fmt.Errorf("%v, please check logs, and retry again.", blockVolumeExpand.ErrMsg)
```
Comment (phlogistonjohn) [inline executors/cmdexec/block_volume.go:362]: The wording here strikes me as a bit odd. Perhaps:
```
Failed to expand block volume: %v (see logs for details, and retry the operation)
&
Failed to expand block volume (see logs for details, and retry the operation)
```

File: apps/glusterfs/operations_block_volume.go (hunk 1 above)
Comment (pkalever) [reply apps/glusterfs/operations_block_volume.go]: ok

File: executors/cmdexec/block_volume.go (hunk 2 above)
Comment (pkalever) [reply executors/cmdexec/block_volume.go:300]: ok

File: executors/cmdexec/block_volume.go (hunk 3 above)
Comment (pkalever) [reply executors/cmdexec/block_volume.go]: ok

**Comment by pkalever** [comment]
@phlogistonjohn @raghavendra-talur update the review comments and tested, it works like this as expected:

```
$ heketi-cli blockvolume expand d2a9ffd1460ffbeaa6c9012f3f3e62e1 --new-size=4
Error: Failed to expand block volume: block volume resize failed: 192.168.121.89:[2.0 GiB] 192.168.121.217:[3.0 GiB] (see logs for details, and retry the operation)

$ heketi-cli blockvolume info d2a9ffd1460ffbeaa6c9012f3f3e62e1
Name: blockvol_d2a9ffd1460ffbeaa6c9012f3f3e62e1
Size: 4
UsableSize: 2
Volume Id: d2a9ffd1460ffbeaa6c9012f3f3e62e1
Cluster Id: c42c101e265dfe42dd8fdcbd02e24ac2
Hosts: [192.168.121.89 192.168.121.106 192.168.121.217]
IQN: iqn.2016-12.org.gluster-block:9cb0013f-fddb-4fde-b8c1-4140bb881240
LUN: 0
Hacount: 3
Username:
Password:
Block Hosting Volume: 24642b23fae92711216840c48a5397d5



Logs:
[heketi] INFO 2020/05/05 21:47:08 Re-executing block volume expansion on [d2a9ffd1460ffbeaa6c9012f3f3e62e1]: volume not fully expanded on backend
[negroni] 2020-05-05T21:47:08+05:30 | 202 |      8.527164ms | localhost:8080 | POST /blockvolumes/d2a9ffd1460ffbeaa6c9012f3f3e62e1/expand
[asynchttp] INFO 2020/05/05 21:47:08 Started job b8518eed6450649750586410ae353d96
[heketi] INFO 2020/05/05 21:47:08 Started async operation: Expand Block Volume
[heketi] INFO 2020/05/05 21:47:08 Trying Expand Block Volume (attempt #1/1)

```


Thanks!

**Comment by pkalever** [comment]
FYI: just rebased on master

File: apps/glusterfs/operations_block_volume.go [hunk 4]
```diff
@@ -486,7 +488,10 @@ func (bve *BlockVolumeExpandOperation) Build() error {
 		}
 
 		requiredFreeSize := bve.newSize - bv.Info.Size
-		if bhv.Info.BlockInfo.FreeSize < requiredFreeSize {
+		if requiredFreeSize == 0 {
+			logger.Info("Re-executing block volume expansion on [%v]: volume not fully expanded on backend", bve.bvolId)
```
Comment (raghavendra-talur) [inline apps/glusterfs/operations_block_volume.go:492]: This log line might give an illusion to the user that there is some kind of retry mechanism. My suggestion would be something like:

"executing block volume expand on [%v] as usable size is not same as the size."

But it is not a hard requirement.

File: apps/glusterfs/operations_block_volume.go [hunk 5]
```diff
@@ -647,6 +655,11 @@ func (bve *BlockVolumeExpandOperation) CleanDone() error {
 		} else {
 			logger.Info("Actual Size matches newly requested Size %v", bve.reclaim.actualSize)
 			bv.Info.Size = bve.newSize
+
+			if bve.reclaim.usableSize > 0 {
```
Comment (raghavendra-talur) [inline apps/glusterfs/operations_block_volume.go:659]: Is this check required because heketi might be talking to an older gluster-block that does not have the usableSize field in info output? If so, can we please have that as a comment here?

File: apps/glusterfs/operations_block_volume.go (hunk 1 above)
Comment (raghavendra-talur) [reply apps/glusterfs/operations_block_volume.go]: Oh there was a thread about this log message already.

I replied as a separate comment that this log line might give an illusion to the user that there is some kind of retry mechanism. My suggestion would be something like:

"executing block volume expand on [%v] as usable size is not same as the size."

File: apps/glusterfs/operations_block_volume.go (hunk 4 above)
Comment (pkalever) [reply apps/glusterfs/operations_block_volume.go:492]: I initially had:

```
logger.Info("Attempting to fix failures seen by recent expand operation")
```
The current one is something that @phlogistonjohn suggested.

I personally don't have a preference for this, let me know which one should we go with, from all these 3 type of msgs.

File: apps/glusterfs/operations_block_volume.go (hunk 5 above)
Comment (pkalever) [reply apps/glusterfs/operations_block_volume.go:659]: make sense, will this be fine?

```
 // Older versions of gluster-block might not have a field in the         
 // `# gluster-block info ...` output from which we calculate usableSize. 
 // In such cases reclaim.usableSize will be set to zero.                 
 if bve.reclaim.usableSize > 0 {                                          
   bv.Info.UsableSize = bve.reclaim.usableSize                            
   logger.Info("Usable Size is: %v", bv.Info.UsableSize)                  
 }                                                                        
```

Comment (raghavendra-talur) [reply apps/glusterfs/operations_block_volume.go:659]: Yes, that looks good.

File: apps/glusterfs/operations_block_volume.go (hunk 4 above)
Comment (raghavendra-talur) [reply apps/glusterfs/operations_block_volume.go:492]: @phlogistonjohn which one do you prefer?

Comment (phlogistonjohn) [reply apps/glusterfs/operations_block_volume.go:492]: I would keep "Re-executing" and swap out the stuff after the brackets for something like @raghavendra-talur has suggested. It's the most precise statement: "Re-executing block volume expansion on [%v]: usable size is not same as the size`
`

Comment (phlogistonjohn) [reply apps/glusterfs/operations_block_volume.go:492]: The reason for "Re-executing" is that its a breadcrumb/reminder that Build has basically already done it's thing and nothing new needs to be built but Exec will be redoing what was already attempeted.

Comment (pkalever) [reply apps/glusterfs/operations_block_volume.go]: done :+1:

File: apps/glusterfs/operations_block_volume.go (hunk 5 above)
Comment (pkalever) [reply apps/glusterfs/operations_block_volume.go:662]: cool

**Comment by pkalever** [comment]
Addressed the review comments from @raghavendra-talur . Thanks!
